### DISCUSSION TITLE
Problems with Definition search

### DISCUSSION THREAD

**Discussion by shusterboris** [oracle/opengrok#4379] [Q&A] [answered]
I need search results by definition to show not only function definitions, but also their declarations in h-files(prototypes) for C language. It worked in old versions, but for OpenGrok 1.8 by default this feature is off. Vladak quite rightly told me that these are the ctags settings. I found a bad way to solve this problem :-). I replaced line 171 in Ctags.java from command.add("--kinds-c=**+l**"); to command.add("--kinds-c=+**lpx**"); 
Maybe someone can suggest how to achieve the same result using the options file for ctags. 
And, the second, search results after changing the code not only corrects. As you can see, second line in results defines as prototype 
![image](https://github.com/oracle/opengrok/assets/25655449/96147fe0-a99c-432d-bf23-6d0b42c7172d)
But it is definition.
![image](https://github.com/oracle/opengrok/assets/25655449/a7e439a6-6367-49dc-ac6d-75508d414b21)
There are another discrepancies when call of function defines as declaration (prototype)
Would appreciate any advise

**Comment by shusterboris** [comment]
Is it correct? I can create options file something like: 
--regex-asm=/^[ \t]*(ENTRY_NP|ENTRY|RTENTRY)+\(([a-zA-Z0-9_]+)\)/\2/f,function/
--regex-<LANG> is the language for, next part is regular expression, last group is keyword: function, variable, prototype, etc
\1 & \2 ????
Do the syntax documented somewhere.
I think this this option file can solve my question

**Reply by vladak** [reply, marked as answer]
These are Universal ctags options so any documentation would have to be located there.

**Comment by shusterboris** [comment]
Another issue related to this thread. Ok, when the options for tags are set in source code or by universal tag option, the web application shows the characteristics of the string: function, prototype, and so on (in the old version, in the Search object it was called TAG) But the API request http://domain/pai/v1/search?<parameters> does not return the tag field in json result regardless of setting. Do you plan to add? It would be helpful

**Reply by vladak** [reply]
That would be an enhancement, please file one.

**Reply by shusterboris** [reply]
Thank you. But how can I do it? Add to ideas?

**Reply by vladak** [reply]
This is an area of OpenGrok I am not too familiar with. The tags are displayed via `ContextFormatter` used in `UnifiedHighlighter`. Specifically, they are acquired from respective Lucene Document here: https://github.com/oracle/opengrok/blob/ab585479d2e343a8f83301ce5dcff2f7d18d3159/opengrok-indexer/src/main/java/org/opengrok/indexer/search/context/Context.java#L151-L153 and then `ContextFormatter` writes them to the page in https://github.com/oracle/opengrok/blob/ab585479d2e343a8f83301ce5dcff2f7d18d3159/opengrok-indexer/src/main/java/org/opengrok/indexer/search/context/ContextFormatter.java#L343-L349
The `marks` list used to do the matching is built while constructing the highlighting in `ContextFormatter#format()` so another way has to be devised to make this work in `SearchEngine` used in `SearchController` to drive the search RESTful API. A crude way might be to parse the results for the `<b>` HTML tag and extract the symbols and then perform the tag matching. That method will be obliterated if #2612 is implemented, though.

**Reply by vladak** [reply]
Actually, looking at the `results` in `SearchController#search()`, the `tags` field is filled, so this is just a matter of remapping it here: https://github.com/oracle/opengrok/blob/ab585479d2e343a8f83301ce5dcff2f7d18d3159/opengrok-web/src/main/java/org/opengrok/web/api/v1/controller/SearchController.java#L91-L94

**Reply by vladak** [reply]
PR #4391 contains the changes.

**Reply by shusterboris** [reply]
Incredible! I'm amazed at the speed of the reaction
